### PR TITLE
sql/pgwire: fix encoding of int4 and bpchar in tuples 

### LINKED ISSUE #58069: sql/pg_wire: Binary encoding of int4 in tuple doesn't match Postgres
**Describe the problem**

It seems like Cockroach encodes 4-byte ints in a tuple as if they were 8-bytes ints in the binary encoding.

**To Reproduce**

If possible, provide steps to reproduce the behavior:

1. Add ````1::int4```` to the `"(%s,null)"` test in `pkg/cmd/generate-binary/main.go`
2. Regenerate `encodings.json` following instructions from `pkg/sql/pgwire/encoding_test.go`
3. Run `TestEncodings` in `pkg/sql/pgwire/encoding_test.go`
4. See error:

```
=== RUN   TestEncodings/encode/FormatBinary
    encoding_test.go:166: unexpected binary encoding:
        	[0 0 0 2 0 0 0 20 0 0 0 8 0 0 0 0 0 0 0 1 0 0 2 193 255 255 255 255] found,
        	[0 0 0 2 0 0 0 23 0 0 0 4 0 0 0 1 0 0 2 193 255 255 255 255] expected
        --- FAIL: TestEncodings/encode/FormatBinary (0.00s)
```

**Expected behavior**

After adding this new test case, the test should pass.

**Additional data / screenshots**

**Environment:**

 - CockroachDB version: CCL v21.1.0-alpha.00000000-660-g6559138f9e
 - Server OS: MacOS 10.15.7
 - Client app: N/A

**Additional context**

CC @rafiss

### PR DESCRIPTION
**Pull request by rafiss**
fixes #58069

This includes 3 commits:

### sql/pgwire: make PGTest pass against PostgresSQL

There are a few minor differences: dataTypeSize is different for tuples,
and PG does not show seconds offset for times if it is zero.

### sql/pgwire: fix encoding of collated strings

Release note (bug fix): Integers inside of tuples were not being encoded
properly when using the binary format for retrieving data. This is now
fixed, and the proper integer width is reported.

Release note (bug fix): Blank-padded chars (e.g. CHAR(3)) were not being
encoded correctly when returning results to the client. Now they
correctly include blank-padding when appropriate.

Release note (bug fix): Collated strings were not encoded with the
proper type OID when sending results to the client if the OID was
for the `char` type. This is now fixed.

### CODE REVIEW

File: pkg/sql/pgwire/types.go [hunk 1]
```diff
@@ -422,7 +410,7 @@ func (b *writeBuffer) writeBinaryDatum(
 		b.writeLengthPrefixedString(v.LogicalRep)
 
 	case *tree.DString:
-		b.writeLengthPrefixedString(resolveBlankPaddedChar(string(*v), t))
+		b.writeLengthPrefixedString(tree.ResolveBlankPaddedChar(string(*v), t))
 
 	case *tree.DCollatedString:
 		b.writeLengthPrefixedString(v.Contents)
```
Comment (otan) [inline pkg/sql/pgwire/types.go:416]: do we need to resolve blank padded chars for collated strings too?

Comment (rafiss) [reply pkg/sql/pgwire/types.go:416]: yeah i think i should add it

File: pkg/sql/pgwire/encoding_test.go [hunk 2]
```diff
@@ -184,6 +184,11 @@ func TestEncodings(t *testing.T) {
 			case *tree.DTuple:
 				// Unsupported.
 				continue
+			case *tree.DCollatedString:
+				// Decoding collated strings is unsupported by this test. The encoded
```
Comment (otan) [inline pkg/sql/pgwire/encoding_test.go:188]: treating collated strings differently to postgres has truly been a pain :(

**Review by otan — APPROVED** [review]
nice. are we backporting this?

File: pkg/sql/pgwire/encoding_test.go (hunk 2 above)
Comment (rafiss) [reply pkg/sql/pgwire/encoding_test.go:188]: yeah the gift that keeps giving

**Comment by rafiss** [comment]
tftr!

bors r=otan

**Comment by rafiss** [comment]
bors r=otan

**Pull request merged**